File: 23.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        def mergeList(a: ListNode, b: ListNode):
            dummy = ListNode(-1)
            rhead = dummy
            while a and b:
                if a.val < b.val:
                    rhead.next = a
                    a = a.next
                else:
                    rhead.next = b
                    b = b.next
                rhead = rhead.next
            rhead.next = a or b
            return dummy.next
        if not lists: return None
        n = len(lists)
        while (n > 1):
            halfLen = int((n + 1) / 2)
            # print(halfLen)
            for i in range(int(n/2)):
                # print(i, i + halfLen)
                lists[i] = mergeList(lists[i], lists[i + halfLen])
            n = halfLen
        return lists[0]
    

File: 23.cpp
#include <iostream>
#include <vector>

using namespace std;

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if (lists.empty()) return NULL;
        int len = lists.size();
        while (len > 1)
        {
            int halfLen = (len + 1) / 2;
            for(int i = 0; i < len / 2; ++i) {
                lists[i] = mergeTwoList(lists[i], lists[i+halfLen]);
            }
            len = halfLen;
        }
        return lists[0];
    }

    ListNode* mergeTwoList(ListNode* a, ListNode* b) {
        ListNode *dummy = new ListNode(-1);
        ListNode *rhead = dummy;
        while (a and b) {
            if (a->val <= b->val) {
                rhead->next = a;
                a = a->next;
            }else {
                rhead->next = b;
                b = b->next;
            }
            rhead = rhead->next;
        }
        rhead->next = a ? a : b;
        return dummy->next;
    }
};